### PR TITLE
os/bluestore:  set bluestore_min_alloc_size to 4K for spinners

### PR DESCRIPTION
**Pull request by ifed01**
This patch definitely has some impact of BlueStore read-write performance.
One can find intermediate and final benchmark results (a bit unsorted) under the following link :
https://docs.google.com/spreadsheets/d/1DzTEumM2hCnUu9Ceib7GfM035zFmPgPHmXkeTW5t6pA/edit?usp=sharing

Signed-off-by: Igor Fedotov <ifedotov@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by markhpc** [comment]
@ifed01 The intent here is that we are also making the hybrid allocator and your deferred write PR default right?

**Comment by ifed01** [comment]
> @ifed01 The intent here is that we are also making the hybrid allocator and your deferred write PR default right?

Missed switching to hybrid allocator - fixed now.
Big deferred writes are enabled by default

**Comment by jdurgin** [comment]
In terms of testing, qa/objectstore has explicitly set allocators, so we should add one for hybrid there (the compression-based ones use the default allocator, but we'll want non-compressed coverage too).

Is there any resaon to use the avl allocator when hybrid is available? Perhaps we could replace it in the test suite with hybrid.

**Comment by jdurgin** [comment]
another possibility is adding a way to randomly choose the allocator, like we had for the messenger with ``ms type = random``, to avoid too much test suite expansion.

**Comment by ifed01** [comment]
> In terms of testing, qa/objectstore has explicitly set allocators, so we should add one for hybrid there (the compression-based ones use the default allocator, but we'll want non-compressed coverage too).
> 
> Is there any resaon to use the avl allocator when hybrid is available? Perhaps we could replace it in the test suite with hybrid.
Good catch, done! Thanks!

**Comment by ifed01** [comment]
> another possibility is adding a way to randomly choose the allocator, like we had for the messenger with `ms type = random`, to avoid too much test suite expansion.

Personally I prefer non-random test cases most of time. But I realize such a randomization is a trade-off between coverage and execution time. So probably it makes sense but let's do that later.

**Comment by ifed01** [comment]
> @ifed01 The unittest_bluefs failure in make check seems related. Could you please add a link to your performance investigation or even a summary of it in this PR for future reference.

Fixed, thanks!

**Comment by jdurgin** [comment]
> > another possibility is adding a way to randomly choose the allocator, like we had for the messenger with `ms type = random`, to avoid too much test suite expansion.
> 
> Personally I prefer non-random test cases most of time. But I realize such a randomization is a trade-off between coverage and execution time. So probably it makes sense but let's do that later.

Sounds good

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2020-04-26_09:25:11-rados-wip-kefu-testing-2020-04-26-1313-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-04-26_13:55:50-rados-wip-kefu-testing-2020-04-26-1314-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-04-27_01:47:10-rados-wip-kefu-testing-2020-04-26-1314-distro-basic-smithi/

**Comment by smithfarm** [comment]
Is this a candidate for backport to octopus and/or nautilus?

**Comment by disaster123** [comment]
@jdurgin i've seen that the hybrid allocator was backported to nautilus and is already merged. but those changes here regarding 4K for spinners isn't. Is this expected? Which severeal large changes do you refer to?

**Comment by disaster123** [comment]
@ifed01 @jdurgin @tchaikov  i've seen that the hybrid allocator was backported to nautilus and is already merged. but those changes here regarding 4K for spinners isn't. Is this expected?

**Comment by nh2** [comment]
This wasn't mentioned in any changelog.

Also this man page is now outdated: https://github.com/ceph/ceph/blame/a2f5a3c1dbfa4dce41e25da4f029a8fdb8c8d864/doc/man/8/rbd.rst#L830

**Comment by nh2** [comment]
Also, it would be nice if https://docs.ceph.com could mention `bluestore_min_alloc_size` at all!

File: src/common/options.cc [hunk 1]
```diff
@@ -4251,7 +4251,7 @@ std::vector<Option> get_global_options() {
     .set_long_description("Chunks larger than this are broken into smaller chunks before being compressed"),
 
     Option("bluestore_compression_max_blob_size_hdd", Option::TYPE_SIZE, Option::LEVEL_ADVANCED)
-    .set_default(512_K)
+    .set_default(64_K)
```
Comment (nh2) [inline src/common/options.cc:4254]: @ifed01 What informs/informed the change to the compression max block sizes, e.g. `bluestore_compression_max_blob_size_hdd` from 512K to 64K?

On ZFS, the size of the blob apparently has a massive impact on compression ratio for e.g. `zstd`; quoting from there:

```
recordsize  compression ratio
1M          2.63x
512K        2.24x
256K        1.72x
64K         1.03x
```

While for `lz4` it it didn't make a big difference:

https://github.com/openzfs/zfs/discussions/11293#discussioncomment-232934

Comment (nh2) [reply src/common/options.cc:4254]: I filed an issue for that:

https://tracker.ceph.com/issues/74558

Comment (nh2) [reply src/common/options.cc:4254]: To articulate a bit better:

I don't understand why the *max* blob size is reduced in a PR whose purpose is to reduce a *min* alloc size.

Shouldn't it leave the max nice and large?

Generally what would be the drawback of leaving `bluestore_compression_max_blob_size` very large? Neither the docs nor the PR seem to say.

**Pull request merged**